bundles/org.openhab.binding.bluetooth.bluegiga/src/main/java/org/openhab/binding/bluetooth/bluegiga/internal/BlueGigaSerialHandler.java

```diff
@@ -474,26 +230,74 @@ private void checkIfAlive() {
         }
     }
 
-    /**
-     * Notify handler event listeners that the handler was bluegigaClosed due to an error specified as an argument.
-     *
-     * @param reason the reason to bluegigaClosed
-     */
-    private void notifyClosed(Exception reason) {
-        // It should be safe enough not to use the NotificationService as this is a fatal error, no any further actions
-        // can be done with the handler, a new handler should be re-created
-        // There is another reason why NotificationService can't be used - the listeners should be notified immediately
-        for (BlueGigaHandlerListener listener : handlerListeners) {
-            try {
-                listener.bluegigaClosed(reason);
-            } catch (Exception ex) {
-                logger.warn("Execution error of a BlueGigaHandlerListener listener.", ex);
-            }
-        }
-    }
+    private Thread createBlueGigaBLEHandler() {
+        final int framecheckParams[] = new int[] { 0x00, 0x7F, 0xC0, 0xF8, 0xE0 };
+        return new Thread("BlueGigaBLEHandler") {
+            @Override
+            public void run() {
+                int exceptionCnt = 0;
+                logger.trace("BlueGiga BLE thread started");
+                int[] inputBuffer = new int[BLE_MAX_LENGTH];
+                int inputCount = 0;
+                int inputLength = 0;
 
-    interface BluetoothListener<T extends BlueGigaResponse> {
-        boolean transactionEvent(BlueGigaResponse response);
-    }
+                while (!close) {
+                    try {
+                        int val = inputStream.read();
+                        if (val == -1) {
+                            continue;
+                        }
 
+                        inputBuffer[inputCount++] = val;
+
+                        if (inputCount < 4) {
+                            // The BGAPI protocol has no packet framing, and no error detection, so we do a few
+                            // sanity checks on the header to try and allow resyncronisation should there be an
+                            // error.
+                            // Byte 0: Check technology type is bluetooth and high length is 0
+                            // Byte 1: Check length is less than 64 bytes
+                            // Byte 2: Check class ID is less than 8
+                            // Byte 3: Check command ID is less than 16
+                            if ((val & framecheckParams[inputCount]) != 0) {
+                                logger.debug("BlueGiga framing error byte {} = {}", inputCount, val);
+                                inputCount = 0;
+                                continue;
+                            }
```

Comment:
I don't know how well this works, but I think of much more reliable (albeit more complicated) ways of checking the frame header. Might not be worth addressing now but should probably be noted for later.

Reply:
Ah, just saw you mentioned this exact piece of code above. In that case, I would suggest changing this code to keep track of a 4 byte window which you shift the buffer one byte at a time. Using a ByteBuffer might be worth looking into.

Reply:
Actually better yet, you could wrap the InputStream in a BufferedInputStream and use its mark/reset functionality.

Reply:
Following code is not very readable, especially as inputCount is increased before this check. Anyhow, I have invented this, but moved the code to own method as earlier all code was in directly in the constructor.

Reply:
I realize that you didn't write the code originally, but the whole point of this PR is to fix a many issues as we can with the bluegiga binding no?

Reply:
Well, yes and no. I definitely like to fix the issues, but don't want to rewrite the whole binding. I have too many other OH related staff to do and this bluetooth thing was just a side track for me. I was really close to write just simple python script to read my one bluetooth device, but then decided to implement the binding for it as it took only one day to implement. But now I have already used three weeks my spare time to fix bluegiga binding itself.

Reply:
I completely understand. Lets just sort out the name discovery issue then call this PR done.
The other stuff can be addressed in a later PR (Probably mine).